drivers/gpu/drm/rp1/rp1-vec/rp1_vec_hw.c

```diff
@@ -46,334 +53,371 @@ static const struct rp1vec_ipixfmt my_formats[] = {
+	u16  ref_vfp;		   /* nominal (vsync_start - vdisplay) when max height */
+	bool interlaced;	   /* set for interlaced			       */
+	bool first_field_odd;	   /* depends confusingly on line numbering convention */
+	s16  scale_v;		   /* V scale in 2.8 format (for power-of-2 CIC rates) */
+	s16  scale_u;		   /* U scale in 2.8 format (for power-of-2 CIC rates) */
+	u16  scale_y;		   /* Y scale in 2.8 format (for power-of-2 CIC rates) */
+	u16  de_end;		   /* end of horizontal Data Active period at 108MHz   */
+	u16  de_bgn;		   /* start of horizontal Data Active period	       */
+	u16  half_lines_per_field; /* number of half lines per field		       */
+	s16  pedestal;		   /* pedestal (1024 = 100IRE) including FIR overshoot */
+	u16  scale_luma;	   /* back end luma scaling in 1.15 format wrt DAC FSD */
+	u16  scale_sync;	   /* back end sync scaling / blanking level as above  */
+	u32  scale_burst_chroma;   /* back end { burst, chroma } scaling	       */
+	u32  misc;		   /* Contents of the "EC" register except rate,shift  */
+	u64  nco_freq;		   /* colour carrier frequency * (2**64) / 108MHz      */
+	u32  timing_regs[14];	   /* other back end registers 0x84 .. 0xB8	       */
 };
 
-/* { NTSC, PAL, PAL-M } x { progressive, interlaced } x { 13.5 MHz, 15.428571 MHz } */
-static const struct rp1vec_hwmode rp1vec_hwmodes[3][2][2] = {
+/* { NTSC, PAL, PAL-M } x { progressive, interlaced } */
+static const struct rp1vec_hwmode rp1vec_hwmodes[3][2] = {
 	{
 		/* NTSC */
 		{
-			{
-				.total_cols = 724,
-				.rows_per_field = 240,
-				.ref_hfp = 12,
-				.ref_vfp = 2,
-				.interlaced = false,
-				.first_field_odd = false,
-				.yuv_scaling = 0x1071d0cf,
-				.back_end_regs = {
-					0x039f1a3f, 0x03e10cc6, 0x0d6801fb, 0x023d034c,
-					0x00f80b6d, 0x00000005, 0x0006000b, 0x000c0011,
-					0x000a0106, 0x00000000, 0x00000000, 0x00000000,
-					0x00000000, 0x00170106, 0x00000000, 0x004c020e,
-					0x00000000, 0x007bffff, 0x38518c9a, 0x11195561,
-					0x02000200, 0xc1f07c1f, 0x087c1f07, 0x00000000,
-					0x0be20200, 0x20f0f800, 0x265c7f00, 0x000801ec,
-				},
-			}, {
-				.total_cols = 815,
-				.rows_per_field = 240,
-				.ref_hfp = 16,
-				.ref_vfp = 2,
-				.interlaced = false,
-				.first_field_odd = false,
-				.yuv_scaling = 0x1c131962,
-				.back_end_regs = {
-					0x03ce1a17, 0x03e10cc6, 0x0d6801fb, 0x023d034c,
-					0x00f80b6d, 0x00000005, 0x0006000b, 0x000c0011,
-					0x000a0106, 0x00000000, 0x00000000, 0x00000000,
-					0x00000000, 0x00170106, 0x00000000, 0x004c020e,
-					0x00000000, 0x007bffff, 0x38518c9a, 0x11195561,
-					0x02000200, 0xc1f07c1f, 0x087c1f07, 0x00000000,
-					0x0be20200, 0x20f0f800, 0x265c7f00, 0x000801ac,
-				},
+			.max_rows_per_field = 240,
+			.ref_vfp = 2,
+			.interlaced = false,
+			.first_field_odd = false,
+			.scale_v = 0x0cf,
+			.scale_u = 0x074,
+			.scale_y = 0x107,
+			.de_end	 = 0x1a4f,
+			.de_bgn	 = 0x038f,
+			.half_lines_per_field = 526,
```

Comment:
I might need to change this number to 524, to match the PAL-M version. typical user expectations and indeed the advertised DRM mode. Though sadly that will cause more colour artifacts.

Reply:
Hmm, should we likewise support both 312 and 313 lines for Progressive PAL? I thought, so have changed my mind.
Unlike 263 lines, 313 seems to provide no benefit, and one video converter dislikes it (it displays in black and white).